### DISCUSSION TITLE
Language server cannot infer cross-file types correctly

### DISCUSSION THREAD

**Discussion by Ziqi-Yang** [wjakob/nanobind#673] [Q&A]
I have a question related to generic python stub file and LSP.

Test code:
```py
from llvmpym import core, utils

c = core.Context.get_global_context()

with open("./test0.ll") as f:
    asm_str = f.read()
    m: core.Module = utils.parse_assembly(asm_str)
```
Langue server shows `utils.parse_assembly(asm_str)` returns `unknown_type`, but it should return a `Module` type.

Package structure:
```
.
├── analysis.py
├── bit_reader.py
├── core.py
├── disassembler.py
├── error_handling.py
├── __init__.py
├── llvmpym_ext
│   ├── analysis.pyi
│   ├── bit_reader.pyi
│   ├── core.pyi
│   ├── disassembler.pyi
│   ├── error_handling.pyi
│   ├── __init__.pyi
│   ├── py.typed
│   ├── support.pyi
│   ├── target_machine.pyi
│   ├── target.pyi
│   └── utils.pyi
├── llvmpym_ext.abi3.so
├── support.py
├── target_machine.py
├── target.py
└── utils.py

```
`utils.py`(the content of other `.py` files are similar):
```
from .llvmpym_ext.utils import *
```

`utils.pyi`:
```py
import llvmpym_ext.core

def parse_assembly(isam: str) -> llvmpym_ext.core.Module: ...
```
Language server cannot infer `llvmpym_ext.core.Module` correctly. When editing, the docs for that function generated by language server shows the result type is `unknown` (`basedpyright` and `pyright` are tested).
Nor can the following code work:  
```py
import llvmpym

def parse_assembly(isam: str) -> llvmpym.llvmpym_ext.core.Module: ...
```
Nor can this:
```py
import llvmpym
# import llvmpym_ext.core

def parse_assembly(isam: str) -> llvmpym.core.Module: ...
```
However, language server can get typed information when classes are in the same file like this:
```py
class Argument(Value):
    """Argument"""

    @property
    def parent(self) -> Function:
        """Obtain the function to which this argument belongs."""

class Function(GlobalObject):
    """Function"""
```

**Comment by wjakob** [comment]
Before going all the way an LSP, I would try type-checking with pyright or mypy. (Or even both, since they generate different errors that are often quite useful to track down problems). Likely there is some fault with the stubs that causes the LSP to fall back to `Unknown`, but the type-checkers will tell you why. You can also directly type-check the stubs themselves.
